### DISCUSSION TITLE
Transforming a subset of a list while returning full list?

### DISCUSSION THREAD

**Discussion by skokenes** [duckdb/duckdb#6683] [Ideas]
With bitstrings, you can use `set_bit` to change a specific bit while returning the modified bitstring at the end, like:
```sql
set_bit('0110010'::BIT, 2, 0) -- returns '0100010'
```
This is very useful! It would be equally useful in data analysis tasks to be able to do the same with lists. Today we have the `list_transform` function which will run lambda for _every_ element of a list, like:
```sql
-- l is [4,5,6]
list_transform(l, x -> x + 1) -- returns [5, 6, 7]
```

But what I need to update just a specific member of the list, like say increment the second element by 1? To accomplish today, I have to use some convoluted list concatenation to accomplish, like:
```sql
list_concat(l[1:1], [l[2]+1], l[3:NULL]) -- returns [4,6,6]
```

It would be useful to have methods to easily mutate a list at specific indexes while returning the updated list, something like
```sql
list_transformAt(l, [1:2], x -> x+1) -- returns [5,6,6]
-- OR
list_setAt(l,2,0) -- returns [5,0,6]
```

**Comment by Mytherin** [comment]
You could use a macro to define such a function:

```sql
create macro list_set_at(arr, index, value) as list_concat(list_concat(arr[:index - 1], [value]), arr[index + 1:]);
select list_set_at([4, 5, 6], 2, 0) as l;
┌───────────┐
│     l     │
│  int32[]  │
├───────────┤
│ [4, 0, 6] │
└───────────┘
```

**Comment by skokenes** [comment]
Ok thanks, I will do that. Is that a performant way to do an operation like that? Just wondering if there was "native" support for such a function, would it be more efficient?

**Reply by Mytherin** [reply]
A native version would be faster but it should not be too slow either way.

**Comment by pankajp** [comment]
I came up with a more generalized solution but it doesn't work:
```
CREATE MACRO list_apply_with_idx(li, fn) AS [fn(i, li[i]) for i in range(1, len(li)+1)];
SELECT list_apply_with_idx(li, (i, x) -> CASE WHEN i <= 2 THEN x+1 ELSE x END) FROM (VALUES ([4, 5, 6])) t(li);
```
> duckdb.InvalidInputException: Invalid Input Error: Attempting to execute an unsuccessful or closed pending query result
> Error: Catalog Error: Scalar Function with name fn does not exist!
> Did you mean "ln"?

**Is there a way to pass on a `lambda` to a macro?**

Until then, the following more verbose equivalent form works:
```
SELECT [CASE WHEN i <= 2 THEN li[i]+1 ELSE li[i] END for i in range(1, len(li)+1)] FROM (VALUES ([4, 5, 6])) t(li);
```

**Reply by Mytherin** [reply]
It is not possible to pass in a lambda function to a macro currently.

**Reply by Alex-Monahan** [reply]
I was thinking of using this approach for some SQL Gymnastics, and found somewhat of a workaround. If you explicitly create a function (instead of using a lambda), you can just redefine the meaning of that function immediately before calling your macro. 
```sql
CREATE OR REPLACE FUNCTION fn(i, x) AS ( CASE WHEN i <= 2 THEN x::BIGINT+1 ELSE x::BIGINT END);
CREATE MACRO list_apply_with_idx(li) AS [fn(i, li[i]) for i in range(1, len(li)+1)];
SELECT list_apply_with_idx(li) FROM (VALUES ([4, 5, 6])) t(li);

CREATE OR REPLACE FUNCTION fn(i, x) AS ( CASE WHEN i <= 2 THEN x::BIGINT+100 ELSE x::BIGINT END);
SELECT list_apply_with_idx(li) FROM (VALUES ([4, 5, 6])) t(li);
```

**Reply by pankajp** [reply]
Nice hack, just saw this mentioned in the recent blog post as well :)
